### DISCUSSION TITLE
Exclude a result from a matcher

### DISCUSSION THREAD

**Discussion by ZeeD** [explosion/spaCy#8553] [Help: Coding & Implementations] [answered]
Hi all!

as exposed on [stackoverflow](https://stackoverflow.com/questions/68182366/exclude-a-result-from-a-matcher) I'm trying to write a pattern that let me match a sequence of elements excluding the ones that are preceded by a key.

this is a complete example:

```python
from spacy import load
from spacy.matcher import Matcher

nlp = load('en_core_web_sm')
matcher = Matcher(nlp.vocab, validate=True)
matcher.add('middle', [[
    {'LEMMA': 'exclude', 'OP': '!'},
    {'ENT_TYPE': 'GPE', 'OP': '+'}
]], greedy='FIRST')

doc = nlp('The Territory for the Titles shall be the United States, its territories and possessions, excluding Puerto Rico')
for _, start, end in matcher(doc):
    print(doc[start:end].text)
```

with the current matcher pattern the output is

```
be the United States
Puerto Rico
```

I *understand* why `Puerto Rico` is printed (`Puerto` matches `{'LEMMA': 'exclude', 'OP': '!'}` and `Rico` `{'ENT_TYPE': 'GPE', 'OP': '+'}`, but I'm trying to find a way to write a pattern that can return *only* `the United States`

Is it possible?

**Comment by lll-lll-lll-lll** [comment]
I tried to use Matcher to extract only "the United States", but was unable to do so. sorry
Instead, I wrote a code to extract only "the United States" using the [PhraseMatcher Object](https://spacy.io/api/phrasematcher).
PhraseMatcher objects are faster and easier to search than Matchers, but they cannot represent abstract descriptions like Matchers.

```py
from spacy.matcher import PhraseMatcher
from spacy import load

nlp = load('en_core_web_sm')

# want words
terms = ["the United States", "Puerto Rico"]

matcher = PhraseMatcher(nlp.vocab,  validate=True)

# Create a Doc Object for each element in the list .  
# If you don't do this, the add function will throw an error.
# See the documentation for details. 
patterns = [nlp.make_doc(text) for text in terms]

matcher.add("America", patterns)

doc = nlp('The Territory for the Titles shall be the United States, its territories and possessions, excluding Puerto Rico')

m = matcher(doc)



for match_id, start, end in m:
    span = doc[start:end]
    print(span.text)

```

output 

```md
the United States
Puerto Rico
```

**Reply by ZeeD** [reply]
Hi!
thanks for the reply.
Unfortunately it seems to me that the PhraseMatcher is not ideal for my usecase: I'm trying to write a rule that could deal with a lot of similar texts, but with different countries. the United States and Puerto Rico are just examples

**Comment by polm** [comment]
The reason this doesn't work is that "Puerto Rico" matches your pattern.

- "Puerto" does not have the lemma "exclude"
- "Rico" is a GPE

This is definitely confusing but it's a straightforward result of your rule unfortunately. 

Generally you should only use the Matcher if it makes things easier, and if you have trouble designing a rule it's easier to fall back to post-processing the results with a function. So in this case just get all GPEs (which doesn't even need a Matcher) and filter them like so:

```
keep = []
for ent in doc.ents:
    if ent.ent_type_ != "GPE": continue
    if ent.start > 0 and doc[ent.start - 1].lemma_ == "exclude": continue
    keep.append(ent)
```

**Reply by ZeeD** [reply]
While I understand your point, I still prefer to work on the patterns instead of manually post processing the results.
My goal is to have a "corpus" of "generic" rules that can extract the desired infos, and return the data "a clean as possible" and to have the post processing step just as a way to "format the output"

**Comment by ZeeD** [comment, marked as answer]
I think I found a way to write a pattern that correctly exclude the "excluded" `GPE`: as written in the stackoverflow answer, the trick I use is to search for something that is both "not the `lemma` exclude" and "not a `GPE`".

the working example is:


```python
from spacy import load
from spacy.matcher import Matcher

nlp = load('en_core_web_sm')
matcher = Matcher(nlp.vocab, validate=True)
matcher.add('territories', [[
    {'LEMMA': {'NOT_IN': ['exclude']},
     'ENT_TYPE': {'NOT_IN': ['GPE']}},
    {'ENT_TYPE': 'GPE', 'OP': '+'}
]], greedy='FIRST')

doc = nlp('The Territory for the Titles shall be the United States, its territories and possessions, excluding Puerto Rico')
for _, start, end in matcher(doc):
    print(doc[start:end].text)
```

return as output just

```
be the United States
```


so, for my goals I am quite happy.

Apart from that (but I think this is another topic) I was wondering if there is / will be a way to specify "capturing groups" in patterns or some other mechanism to say "this piece of pattern is actually interesting" and "this piece of pattern is needed to specify the information I want, but I *don't* want it in the results" (in my case I don't really want `be` but just `the United States`)

**Reply by polm** [reply]
Glad you found a solution! I would re-iterate that while the Matcher is powerful, if you find it difficult to express something, it's often faster from a development perspective to use the Matcher for initial filtering and clean up details with post-processing in normal code rather than getting the perfect pattern.

> Apart from that (but I think this is another topic) I was wondering if there is / will be a way to specify "capturing groups" in patterns or some other mechanism to say "this piece of pattern is actually interesting" and "this piece of pattern is needed to specify the information I want, but I don't want it in the results" (in my case I don't really want be but just the United States)

It's not exactly capturing groups but there is a related feature in #7321, which will be released in v3.1 soon. You may also want to see the older thread #3275 on the topic.
